File: backend/src/main/java/django/Signing.java
package django;

import java.io.IOException;
import java.util.zip.Inflater;

import com.google.common.io.BaseEncoding;
import okio.Buffer;
import okio.InflaterSource;

public class Signing {
    static final BaseEncoding BASE64 = BaseEncoding.base64Url().omitPadding();

    public static <T> T loadString(String s, String key, String salt, Serializer<T> serializer, int maxAge) throws IOException {
        String base64d = new TimestampSigner(key, salt).unsign(s, maxAge);
        boolean decompress = false;
        if (base64d.startsWith(".")) {
            base64d = base64d.substring(1);
            decompress = true;
        }
        byte[] data = BASE64.decode(base64d);
        if (decompress) {
            data = decompress(data);
        }
        return serializer.deserialize(data);
    }

    private static byte[] decompress(byte[] compressed) throws IOException {
        Buffer buffer = new Buffer().write(compressed);
        Inflater inflater = new Inflater();
        Buffer decompressed = new Buffer();
        decompressed.writeAll(new InflaterSource(buffer, inflater));
        return decompressed.readByteArray();
    }

    public interface Serializer<T> {
        T deserialize(byte[] data) throws IOException;
    }
}

File: backend-ts/src/auth/django/signing.ts
import { TimestampSigner } from "./timestamp_signer"
import base64url from "base64url"
import * as zlib from "zlib"

interface Serializer<T> {
  serialize(obj: T): Buffer
  deserialize(data: Buffer): T
}

export class Signing {
  static loadString<T>(s: string, key: string, salt: string, serializer: Serializer<T>, maxAgeMs: number | null): T {
    let base64d = new TimestampSigner(key, salt).unsign(s, maxAgeMs)
    let decompress = false
    if (base64d.startsWith(".")) {
      base64d = base64d.substring(1)
      decompress = true
    }
    let data = base64url.toBuffer(base64d)
    if (decompress) {
      data = this.decompress(data)
    }
    return serializer.deserialize(data)
  }

  static toString<T>(obj: T, key: string, salt: string, serializer: Serializer<T>, compress: boolean): string {
    let data = serializer.serialize(obj)
    let isCompressed = false
    if (compress) {
      let compressed = this.compress(data)
      if (compressed.length < data.length - 1) {
        data = compressed
        isCompressed = true
      }
    }
    let base64d = base64url.encode(data)
    if (isCompressed) {
      base64d = "." + base64d
    }
    return new TimestampSigner(key, salt).sign(base64d)
  }
  
  private static decompress(compressed: Buffer): Buffer {
    return zlib.inflateSync(compressed)    
  }

  private static compress(data: Buffer): Buffer {
    return zlib.deflateSync(data)
  }
}
